p2p/service.py

```diff
@@ -114,12 +119,26 @@ def get_event_loop(self) -> asyncio.AbstractEventLoop:
     def add_finished_callback(self, finished_callback: Callable[['BaseService'], None]) -> None:
         self._finished_callbacks.append(finished_callback)
 
-    def run_child_service(self, child_service: 'BaseService') -> 'asyncio.Future[Any]':
+    def run_task(self, awaitable: Awaitable[Any]) -> None:
+        """Run the given awaitable in the background.
+
+        The awaitable should return whenever this service's cancel token is triggered.
+
+        If it raises OperationCancelled, that is caught and ignored.
+        """
+        async def f() -> None:
+            try:
+                await awaitable
+            except OperationCancelled:
+                pass
+        self._tasks.add(asyncio.ensure_future(f()))
```

Comment:
I'm not seeing anywhere that we cleanup finished tasks.  Things like `Discovery.mayber_connect_to_more_peers` which run periodically will result in an ever growing memory footprint.  It looks like we need something to do periodic garbage collection for tasks that are complete.

Reply:
That's why I used a `WeakSet` for `self._tasks`; once a task is done there should be no strong reference to it left, and then it is discarded from the set.

Reply:
https://github.com/ethereum/py-evm/pull/1173/files#diff-bea0ad88fecb099146d53835e5f8fc78R35

Reply:
That's fancy!  :+1: